### ISSUE TITLE
eager_load difference between autoloader classic, bug or feature?

### ISSUE THREAD

**Issue by cj** [fxn/zeitwerk#19]
Hi,

I noticed that when I used `config.eager_load=true` or ran the [Jumpstart](https://github.com/excid3/jumpstart/tree/c6c0de0a9493c617297746ebef99a061f05c1518) template in production which overrides the Devise helper https://github.com/excid3/jumpstart/blob/c6c0de0a9493c617297746ebef99a061f05c1518/app/helpers/devise_helper.rb it was not getting used.  If I however switched to `config.autoloader = :classic` everything worked as expected.

Is this a bug or a feature?

#### Step 1:
- rails new myapp -d postgresql -m https://raw.githubusercontent.com/excid3/jumpstart/c6c0de0/template.rb
- cd myapp

#### Step 2:
- foreman start
- visit http://localhost:5000/users/sign_up
- click the Sign up button and you will see a bootstrap error message above the form (which is correct).

#### Step 3:
- kill foreman
- change `config.eager_load = true` (or run in production) and repeat **Step 2**.  You will notice it is no longer the bootstrap error defined here https://github.com/excid3/jumpstart/blob/c6c0de0a9493c617297746ebef99a061f05c1518/app/helpers/devise_helper.rb plus it is never loaded (which is wrong).

#### Step 4:
- kill foreman
- add `config.autoloader = :classic` to config/application.rb
- Repeat **Step 2** and it will be the correct boostrap error message again.

I'm going to also reference https://github.com/excid3/jumpstart/issues/73 and loop in @excid3.

**Comment by fxn** [comment]
I deleted the previous comment. I have been able to reproduce now the steps above and the `DeviseHelper` module from the app gets the autoload set first. Need to investigate with the real thing.

**Comment by excid3** [comment]
Thanks for checkign it out @fxn! My first assumption was that basically that I was doing thing module override wrong, like you mentioned in your previous comment. Then we found it strange that things were different with eager_load on and off since it wasn't consistent. Curious to learn what's going on. 👍

**Comment by fxn** [comment]
I think I have it.

Essentially, what is happening is that eager loading is implemented as a [breadth-first recursion](https://github.com/fxn/zeitwerk/blob/ef2507779bee17ed4fb0aece9eda4fbcf32cd5f9/lib/zeitwerk/loader.rb#L244-L257) calling `require`, and that is not consistent with the logic while setting autoloads that discards late files with the same relative name mimicking `$LOAD_PATH` semantics.

The reason is that by design we use only absolute file names, and both `require` calls succeed because the file names are different, so the one in devise wins because it runs later.

The gist of the problem is shown here:

```
$ cat a/foo.rb
Foo = :a
$ cat b/foo.rb
Foo = :b
$ cat test.rb
require "zeitwerk"

$VERBOSE = nil

lA = Zeitwerk::Loader.new
lA.push_dir("a")
lA.setup

lB = Zeitwerk::Loader.new
lB.push_dir("b")
lB.setup

p Foo

lA.reload
lB.reload

lA.eager_load
lB.eager_load

p Foo
```

The output of the test is

```
:a
:b
```

which is not consistent, it should print `:a` twice.

In the early prototypes of Zeitwerk I maitained a structure with constants to be loaded, and triggered autoloads instead of using `require`, it could be the case that I might need to go back to that initial implementation.

I'll think how to address it, but we'll do something.

**Comment by fxn** [comment]
Could you please try branch `eager-load-skip`? I have drafted a fix.

**Comment by cj** [comment]
@fxn that works for me!

@excid3 did you want to give it a try?

**Comment by excid3** [comment]
Worked perfect. Thanks @fxn, that was quick!

**Comment by cj** [comment]
I think I may have found another "bug", if set `config.eager_load = false` (run it in development), 'foreman start' and then add this line `after_action { pp 'foo' }` to `app/controllers/application_controller.rb`; refresh the page and you will notice `foo` does not get logged.  If you repeat these steps (delete that line added to application.rb first) and before running `foreman start` set `config.autoloader = :classic`, when you add that line in now and refresh the page it will print out `foo` just fine.

**Comment by fxn** [comment]
Regarding eager loading, I'll polish the code and publish a new version of the gem.

The issue @cj mentions seems different (and surprising!), maybe deserves a different ticket. Did you restart the server after changing the configuration? If yes, could you please enable logging in `config/application.rb`:

```
Rails.autoloaders.logger = method(:puts)
```

reproduce, and share the log specifying at which point did you change and reload?

**Comment by fxn** [comment]
@cj just in case... do you have `config.cache_classes` set to `false`? It should be for reloading to work.

**Comment by cj** [comment]
@fxn I can open up a new ticket if you would like.

I'm using the same template as used in this issue which is `rails new myapp -d postgresql -m https://raw.githubusercontent.com/excid3/jumpstart/c6c0de0/template.rb` it creates a standard `config/environments/development.rb` file with the contents of https://gist.github.com/cj/45d48da8b7bfbd11bbaaacb04abd6ff8 where `config.cache_classes` is indeed false.

When I say, "it does not get logged", I mean it does not get printed out in the console where the command `foreman start` was ran.

I am adding this line `after_action { pp 'foo' }` to `app/controllers/application_controller.rb` after the rails server has started and finished loading from running `foreman start`.  When I refresh the browser I should see `foo` printed to the console where foreman is running, but I do not.  If however I kill foreman, remove the line I added ``after_action { pp 'foo' }`` and add `config.autoloader = :classic` to `config/environments/development.rb`, start the server back up with `foreman start` and add back in the line `after_action { pp 'foo' }` to `app/controllers/application_controller.rb`; refresh the browser, I see `foo` printed out to the console as expected.

I hope this makes sense, if you need me to record a screencast or go in to more detail let me know.

**Comment by cj** [comment]
One other thing to note is I have spring disabled (http://fuzzyblog.io/blog/rails/2017/03/20/disabling-spring-in-rails.html).

**Comment by fxn** [comment]
@cj did you see the comment above re printing Zeitwerk traces?

Yeah, better create a new ticket for this one please.

**Comment by fxn** [comment]
The just published 1.3.4 has this solved, thanks for the bug report!
